src/shelly_hap_rgb.cpp

```diff
@@ -212,38 +211,84 @@ void RGBWLight::HSVtoRGBW(const HSV &hsv, RGBW &rgbw) const {
   }
 }
 
-void RGBWLight::SetOutputState(const char *source) {
-  LOG(LL_INFO,
-      ("state: %s, brightness: %i, hue: %i, saturation: %i", OnOff(cfg_->state),
-       cfg_->brightness, cfg_->hue, cfg_->saturation));
+void RGBWLight::TurnOn(const std::string &source) {
+  UpdateOnOff(true, source);
+}
 
-  RGBW rgbw{};
+void RGBWLight::TurnOff(const std::string &source) {
+  UpdateOnOff(false, source);
+}
 
-  if (cfg_->state) {
-    HSV hsv{.h = cfg_->hue / 360.0f,
-            .s = cfg_->saturation / 100.0f,
-            .v = cfg_->brightness / 100.0f};
+void RGBWLight::Toggle(const std::string &source) {
+  UpdateOnOff(IsOff(), source);
+}
 
-    HSVtoRGBW(hsv, rgbw);
+bool RGBWLight::IsOn() const {
+  return cfg_->state != 0;
```

Comment:
cfg_->state is a bool, it should be used as bool.

come to think of it - why do we even need separate state when we have brightness?

Reply:
brightness and state (aka On char) are different thinks in hk, a lamp can be off and brightness 100% at the same time.

Reply:
ok, so when you set it to on it goes to the configured brightness. makes sense.

Reply:
> cfg_->state is a bool, it should be used as bool.
> 
> come to think of it - why do we even need separate state when we have brightness?

cfg_->state is declared as an int currently, that's the reason why I did that. Casting int's implicit to bool with !var is an ugly thing from the C century. More modern languages does even not allow that. The other thing is why is it a int currently? If it would be a real bool than this != 0 cast would not be needed.

As Timo said, on/off and brightness are different things in HK, you can change the brightness without turning on the light. if you turn on the light after, it will ramp to the previously set brightness.

Reply:
brightness and state (aka On char) are different thinks in hk, a lamp can be off and brightness 100% at the same time, but led state should be off then.

Reply:
Why do you think state is a int and not a Boolean?
```
void mgos_config_sw_set_defaults(struct mgos_config_sw *cfg) {
  cfg->name = NULL;
  cfg->enable = true;
  cfg->out_inverted = false;
  cfg->in_mode = 1;
  cfg->in_inverted = false;
  cfg->state = false;
  cfg->svc_type = 0;
  cfg->valve_type = 1;
  cfg->initial_state = 3;
  cfg->auto_off = false;
  cfg->auto_off_delay = 0.0;
  cfg->state_led_en = -1;
  cfg->persist_state = false;
}
```
Form `gen/mgos_config.c`, or did I miss something?

Reply:
@jobl1978 
> cfg_->state is declared as an int currently

```
- ["lb.state", "b", false, {title: "State of the light"}]
```

looks like a bool to me. am i looking at the wrong struct?

Reply:
oh, i see. the underlying value for config bools is an int, historically. ok, never mind.

Reply:
Looks like a Boole is a int under the hood.
```
/* sw type struct mgos_config_sw */
struct mgos_config_sw {
  const char * name;
  int enable;
  int out_inverted;
  int in_mode;
  int in_inverted;
  int state;
  int svc_type;
  int valve_type;
  int initial_state;
  int auto_off;
  double auto_off_delay;
  int state_led_en;
  int persist_state;
};
```
from `gen/mgos_config.h`

Reply:
yeah, yeah, you're right, i completely forgot about it. originally i made it so i wouldn't need to deal with sizeof(bool) being potentially different, but it seems pretty stable at 1 byte. however, i don't think we'll win any space if we switch, as fields will be 4 byte aligned anyway...

anyhow, i usually trat bool fields a bool in spirit, if not in the letter :)
i.e., i would use cfg_->state in logical context as if it were a bool, without these extra IsOn / IsOff helpers. the less code, the better.